### PR TITLE
🎉 New Destination: Qdrant (Vector Database)

### LINKED ISSUE #29009: Qdrant
https://milvus.io/

Total swag credits for this connector: 75

### PR DESCRIPTION
**Pull request by topefolorunso**
## What
Adding a new vector destination - Qdrant resolves #29009

## Recommended reading order

1. `destination_qdrant/config.py`
2. `destination_qdrant/indexer.py`
3. `destination_qdrant/destination.py`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

See old WIP PR here: https://github.com/airbytehq/airbyte/pull/29895

### CODE REVIEW

**Comment by topefolorunso** [comment]
Qdrant PR here, @flash1293. I will add docs and unit test later today. Looking forward to your review on this one

File: airbyte-integrations/connectors/destination-qdrant/destination_qdrant/indexer.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+            payload = chunk.metadata
+            payload[self.config.text_field] = chunk.page_content
+            entities.append(
+                models.Record(
+                    id=str(uuid.uuid4()),
+                    payload=payload,
+                    vector=chunk.embedding,
+                )
+            )
+        self._client.upload_records(collection_name=self.config.collection, records=entities)
```
Comment (flash1293) [inline airbyte-integrations/connectors/destination-qdrant/destination_qdrant/indexer.py:92]: Could you set `parallel` to 10 here to speed up the upload?

File: airbyte-integrations/connectors/destination-qdrant/icon.svg
Comment (flash1293) [inline airbyte-integrations/connectors/destination-qdrant/icon.svg:1]: The logo should be square, so without the "drant" part.

**Review by flash1293 — COMMENTED** [review]
Thanks for working on this. Left some comments

**Comment by topefolorunso** [comment]
You can review now @flash1293

**Comment by flash1293** [comment]
CI running here: https://github.com/airbytehq/airbyte/pull/30432

**Comment by topefolorunso** [comment]
One more thing - Please confirm that these Milvus files (.gitignore and base_destination_test.py) are supposed to be there. They were from your previous PR draft.

https://github.com/airbytehq/airbyte/pull/30332/files#diff-5593abea0e89d7a53f722856f79c48100693a2a720056749d6e0abb4238e9077

https://github.com/airbytehq/airbyte/pull/30332/files#diff-404db08519b8c53b5b98830d884484229dafce6b77a05ae7ff33b319aae4689c

**Comment by flash1293** [comment]
> The get_collections method doesn't exactly return a list, so I got the string representation instead and checked if the collection name is in it. I hope that is okay.

What do you mean exactly by they don't return a list? I didn't test yet but the types indicate it's a list of object - let's do this right as the string representation can always have false positives. I recommend testing the code locally by setting up a qdrant cluster (easiest way is to get a free cluster in qdrant cloud), and then testing it without the Airbyte platform via `python main.py check --config secrets/config.json` (put the config you want to test in `secrets/config.json`)


> One more thing - Please confirm that these Milvus files (.gitignore and base_destination_test.py) are supposed to be there. They were from your previous PR draft.

No, those can be removed

**Comment by topefolorunso** [comment]
Fixed now, you can review @flash1293

**Comment by flash1293** [comment]
Hi @topefolorunso , fixed 3 problems:
* The return value of post_sync has to be a list of messages, not a single message
* Setting the index type of record_id and stream feels didn't work, the enum has to be used, not a string
* I had to remove the `parallel` flag as it wouldn't run properly:
```
Traceback (most recent call last):
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/forkserver.py", line 274, in main
    code = _serve_one(child_r, fds,
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/forkserver.py", line 313, in _serve_one
    code = spawn._main(child_r, parent_sentinel)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/spawn.py", line 125, in _main
    prepare(preparation_data)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/runpy.py", line 268, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/run.py", line 29, in <module>
    list(dest.write(config, ConfiguredAirbyteCatalog(**catalog), examples_messages_10k_json))
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/destination_qdrant/destination.py", line 43, in write
    yield from writer.write(configured_catalog, input_messages)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/writer.py", line 57, in write
    self.batcher.add(message.record)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/batcher.py", line 24, in add
    self._flush_if_necessary()
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/batcher.py", line 34, in _flush_if_necessary
    self.flush()
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/batcher.py", line 29, in flush
    self.flush_handler(list(self.buffer))
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/writer.py", line 32, in <lambda>
    self.batcher = Batcher(batch_size, lambda batch: self._process_batch(batch))
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/airbyte_cdk/destinations/vector_db_based/writer.py", line 45, in _process_batch
    self.indexer.index(documents, ids_to_delete)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/destination_qdrant/indexer.py", line 107, in index
    self._client.upload_records(collection_name=self.config.collection, records=entities, parallel=4)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/qdrant_client/qdrant_client.py", line 1509, in upload_records
    return self._client.upload_records(
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/qdrant_client/qdrant_remote.py", line 1898, in upload_records
    self._upload_collection(
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/qdrant_client/qdrant_remote.py", line 1881, in _upload_collection
    for _ in pool.unordered_map(batches_iterator, **updater_kwargs):
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/qdrant_client/parallel_processor.py", line 122, in unordered_map
    self.start(**kwargs)
  File "/Users/joereuter/Clones/airbyte/airbyte-integrations/connectors/destination-qdrant/.venv/lib/python3.9/site-packages/qdrant_client/parallel_processor.py", line 117, in start
    process.start()
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/context.py", line 291, in _Popen
    return Popen(process_obj)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/popen_forkserver.py", line 35, in __init__
    super().__init__(process_obj)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/popen_forkserver.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "/Users/joereuter/.pyenv/versions/3.9.11/lib/python3.9/multiprocessing/spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

@kacperlukawski Do you know what the problem with `parallel` is related to? It also happened when I run the connector directly from the command line via `python main.py`. 

I also profiled the indexing of 10k records with a dimensionality of 96 into a local cluster and it takes roughly 30 seconds for me which seems to be slower than expected. It spends ca 30% of the time doing work on CPU, most of which is converting the embedding vector and creating the `Record` instances:
Screenshot 2023-09-15 at 15 30 28

**Comment by flash1293** [comment]
@topefolorunso  Please also add a check that the url starts with `https://` when running on cloud like here: https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-milvus/destination_milvus/indexer.py#L56

**Comment by kacperlukawski** [comment]
@flash1293 Do you experience a problem with `parallel` on Windows?

**Comment by kacperlukawski** [comment]
I still need time to verify if `parallel` might be used in such a scenario. I'll try to recreate that error and provide a fix.

Overall, it looks fine to me! @flash1293 Is there any way we can make sure Pydantic 2 is being used? That should improve the performance.

**Comment by flash1293** [comment]
Thanks @kacperlukawski - It's good to know pydantic 2 would help here - unfortunately that's a whole ball of yarn. We are aware of the need to upgrade here, I will make sure this is bumping the priority.

**Comment by flash1293** [comment]
Overall, this looks good for a first version - we can improve things like parallel loading in subsequent versions.

CI is running here: https://github.com/airbytehq/airbyte/pull/30432

I everything works fine I'm going to merge after another small round of testing.

**Review by flash1293 — APPROVED** [review]
LGTM, will merge once CI problems are resolved

**Comment by topefolorunso** [comment]
Thanks for your support on this @flash1293. Glad we got this merged finally 🎉

**Pull request merged**